Source/Core/Core/PowerPC/Jit64/Jit64_Util.cpp

```diff
@@ -0,0 +1,420 @@
+    auto xv = reg_value.Bind(BindMode::ReadWrite);
+    auto scratch = regs.gpr.Borrow();
+    MOV_sum(32, scratch, addr, xv);
+    SafeWriteRegToReg((X64Reg)xv, scratch, accessSize, 0, CallerSavedRegistersInUse(), flags);
+    return;
+  }
+
+  // No update so we need to sum things in a scratch register. We can't just pass the offset into
+  // SafeWriteRegToReg because that will clobber the address register we pass in
+  auto scratch = regs.gpr.Borrow();
+  MOV_sum(32, scratch, addr, offset);
+  SafeWriteRegToReg(val, scratch, accessSize, 0, CallerSavedRegistersInUse(), flags);
+}
+
+bool Jit64::UnsafeWriteToConstAddress(int accessSize, GPRRegister& value, u32 address, bool swap)
+{
+  // This needs to be a scratch register if it would be clobbered by SwapAndStore (ie: if we don't
+  // have MOVBE)
+  _assert_msg_(DYNAREC, !WriteClobbersRegValue(accessSize, swap) || !value.HasPPCRegister(),
+               "Asked to clobber a PPC register");
+
+  // If we already know the address through constant folding, we can do a
+  // few interesting things:
+
+  // If this is the WPAR address, we can generate optimized FIFO routines
+  if (jit->jo.optimizeGatherPipe && PowerPC::IsOptimizableGatherPipeWrite(address))
+  {
+    value.LoadIfNotImmediate();
+    UnsafeWriteGatherPipe(accessSize, value, swap);
+    return true;
+  }
+
+  // If we know for sure this is RAM, we can just write to it
+  if (PowerPC::IsOptimizableRAMAddress(address))
+  {
+    value.LoadIfNotImmediate();
+    UnsafeWriteToConstRamAddress(accessSize, value, address, swap);
+    return true;
+  }
+
+  return false;
+}
+
+void Jit64::UnsafeWriteToMemory(int accessSize, GPRRegister& value, OpArg target, bool swap)
+{
+  OpArg arg = value;
+  if (arg.IsImm())
+  {
+    arg = arg.AsImm(accessSize);
+    if (swap)
+      arg = arg.SwapImm();
+    MOV(accessSize, target, arg);
+  }
+  else
+  {
+    if (swap)
+      SwapAndStore(accessSize, target, arg.GetSimpleReg());
+    else
+      MOV(accessSize, target, arg);
+  }
+}
+
+void Jit64::UnsafeWriteGatherPipe(int accessSize, GPRRegister& value, bool swap)
+{
+  u32 gather_pipe = (u32)(u64)GPFifo::m_gatherPipe;
+  _assert_msg_(DYNA_REC, gather_pipe <= 0x7FFFFFFF, "Gather pipe not in low 2GB of memory!");
+
+  // Append this write to the pipe
+  auto scratch = regs.gpr.Borrow();
+  MOV(32, R(scratch), M(&GPFifo::m_gatherPipeCount));
+  UnsafeWriteToMemory(accessSize, value, MDisp(scratch, gather_pipe), swap);
+  ADD(32, R(scratch), Imm8(accessSize >> 3));
+  MOV(32, M(&GPFifo::m_gatherPipeCount), R(scratch));
+  jit->js.fifoBytesThisBlock += accessSize >> 3;
+}
+
+void Jit64::UnsafeWriteToConstRamAddress(int accessSize, GPRRegister& value, u32 address, bool swap)
+{
+  // We can avoid a scratch register if the address is in the lower half
+  if (address > 0x80000000)
+  {
+    // Technically the borrow doesn't last long enough, but nothing will
+    // re-allocate it before we're done
```

Comment:
What?  I don't follow... the lifetime of `addr` lasts to the end of the if statement, which should be long enough?